public/js/helpers/zoom_helper.js

```diff
@@ -113,18 +113,15 @@ export default class Zoom {
   // mapKey returns the corresponding map key, if the zoom meets the correct
   // range and position within the limits, else null.
   static mapKey (zoom, limits, scale) {
+    scale = scale || 1
     let lims = tryDecode(limits)
     let decoded = this.decode(zoom, lims)
-    if (decoded.end !== lims.end) return null
-    if (decoded.start === lims.start) return 'all'
-    let keys = Object.keys(zoomMap)
-    for (let idx in keys) {
-      let k = keys[idx]
```

Comment:
These changes break `mapKey`. Please undo the changes made to this function.

Reply:
This the only part whose change breaks my implementation. I am not aware how `null` value returned falls into place because the zoom button doesn't expect it. IMO `all` should be the default and returning `null` complicates things.

Secondly, In the previous implementation the `decoded.start` and `decoded.end` seems to have more significance than just getting the range. I think comparing the range was a better way to implement this `mapKey`.

Lets work on a compromise that works on all cases.

Reply:
No problem. Let's figure it out. What is the behavior that you are expecting?

The zoom buttons correspond to very specific ranges. If none of those ranges are currently displayed, then no zoom button should be selected. There is no default, only specific ranges for which a button should be visually distinguished. 

As an example, compare this PR to alpha by opening a chart, clicking the `year` button, and then grabbing the left zoom handle and dragging the window back to full range. On this PR, the `year` button will remain highlighted, even though it is not a year, and is actually the correct range for the `all` button. On master, when you begin dragging, the `year` button is no longer selected, because it is no longer a window of a year. Also, when you drag back to full range, the `all` button will become selected.

Reply:
Oh yeah. As far as the ranges go, I think it would be fine if the button is selected based on just the range, dropping the *most recent* requirement. It seems like you were trying to do that. The only reason I hesitate at all is because for the url parameter `?zoom=day`, it should display the most recent, but that doesn't necessarily have to match the behavior of the buttons.

Reply:
>There is no default, only specific ranges for which a button should be visually distinguished.

I don't fully agree with the above statement because if it was true, when no zoom button is active there should never be a chart that is visible. Instead when no zoom button is active, the chart whose zoom corresponds to zoomMap key `all` is visible. IMO this introduces inconsistencies as the other button adjacent to zoom nav do not behave in the same way (i.e. their default values are always active). It also becomes hard to describe a chart if some button sets have an active one while others don't.

Concerning the slider drag feature, well I wasn't aware of its existence and thanks for letting me know about it. And because of the huge volume of data set that historical charts typically deal with, I would rather just use the zoom buttons than try to move the slider in order to select a `year` range or any other range. This ranges deals with exact values (i.e start and end values) and honestly I cannot imagine how someone will use the range slide to select a zoom type.

Don't get me wrong, matching the range selected to its matching zoom type is a cool feature which should remain in place but I doubt users who focus on the range slider at the page bottom will ever notice that the active zoom button at the page top has changed.

Reply:
My use case of `mapKey` focuses on obtaining the correct zoom type from an old URL while your use case focuses on the slider drag feature.
I suggest we add variable to the `mapKey` function that differentiate those two use cases so as to resolve this slider feature break.

Reply:
Based on the implementation when `?zoom=day`, It should work seamlessly as when any other correct case scenario is set. Kindly share more on the exact case scenario(s) where `?zoom=day` would be unpredictable.

Reply:
I'm going to use `month` because it's easier to work with. Below are two views of the same chart. 

![image](https://user-images.githubusercontent.com/6109680/58443085-5c733200-80b5-11e9-9be8-98c7c15c9a71.png)

Which view should the user get when they navigate to `/charts?chart=ticket-price&zoom=month`? 

I think most would agree that it's the bottom chart. Then the question is whether the zoom keyword `month` is only associated with the most recent month? 

But again, we can separate the button behavior from the definition of the URL parameter behavior if it improves the experience. 

I'm still not certain what behavior you are expecting that you aren't seeing. I'm not quite sure what you mean when you say 

> obtaining the correct zoom type from an old URL

Reply:
With old URLs, I meant this: https://github.com/decred/dcrdata/pull/1339#discussion_r286082512

As mentioned here let me add a variable to differentiate the two use cases: https://github.com/decred/dcrdata/pull/1339#discussion_r287894900

Reply:
This is a little confusing. The comment you linked to is about the URL parameter of the `/api/charts` API endpoint , not the `/charts` page. The use of `mapKey` is only used to set the `zoom` range parameter the `/charts` page URL, and is not used for the API endpoint. 

I'd also like to swing back to something above

> when no zoom button is active there should never be a chart that is visible. Instead when no zoom button is active, the chart whose zoom corresponds to zoomMap key all is visible. IMO this introduces inconsistencies as the other button adjacent to zoom nav do not behave in the same way (i.e. their default values are always active). It also becomes hard to describe a chart if some button sets have an active one while others don't.

It seems like you are suggesting that we should always have a button selected, even when it does not accurately describe the displayed range, which I vehemently disagree with. But maybe I'm not understanding your position correctly.

Reply:
>This is a little confusing. The comment you linked to is about the URL parameter of the /api/charts API endpoint , not the /charts page. The use of mapKey is only used to set the zoom range parameter the /charts page URL, and is not used for the API endpoint.

But on the user facing end, parameters  used to query `/api/charts`  API endpoint are stored on the browser's  `/charts` page URL. Therefore IMO if a specific chart has been referenced, the `/charts` page URL will be used instead of `/api/charts` API URL. I made an effort to ensure that the referenced chart page URL will show a chart similar to the originally referenced when accessed.

With the slider feature, it makes sense as to why a `null` value might be returned and this is why I will be adding an extra parameter to differentiate the two use cases rather than create a separate implementation of `mapKey`.